README.md
# Information about ipcalc

This is a modern tool to assist in network address calculations for IPv4 and IPv6.
It acts both as a tool to output human readable information about a network or address,
as well as a tool suitable to be used by scripts or other programs.

It supports printing a summary about the provided network address, multiple
command line options per information to be printed, transparent IPv6 support,
and in addition it will use libGeoIP if available to provide geographic information.

The project started as a fork of the ipcalc tool in the Fedora distribution
but it has since then replaced the original tool.


# How to build

Simply type:
```
$ make
```

In order to access geo-information the application depends on the maxmind
database (libmaxminddb). Alternatively it can be built using the older
libgeoip (libgeoip) or with no geo-information whatsoever. The options
can be provided on the makefile via the variables USE_GEOIP (yes/no),
USE_MAXMIND (yes/no). For example the following command will compile
with no support for geo-information.

```
$ make USE_GEOIP=no USE_MAXMIND=no
```


# Examples

## IPv4

```
$ ./ipcalc --all-info 193.92.150.2/24
Address:        193.92.150.2
Network:        193.92.150.0/24
Netmask:        255.255.255.0 = 24
Broadcast:      193.92.150.255
Reverse DNS:    150.92.193.in-addr.arpa.

Address space:  Internet
Address class:  Class C
HostMin:        193.92.150.1
HostMax:        193.92.150.254
Hosts/Net:      254

Country code:   GR
Country:        Greece
```

```
$ ./ipcalc -pnmb --minaddr --maxaddr --geoinfo --addrspace 193.92.150.2/255.255.255.224
NETMASK=255.255.255.224
PREFIX=27
BROADCAST=193.92.150.31
NETWORK=193.92.150.0
MINADDR=193.92.150.1
MAXADDR=193.92.150.30
ADDRSPACE="Internet"
COUNTRY="Greece"
```

## IPv6

```
$ ./ipcalc --all-info 2a03:2880:20:4f06:face:b00c:0:14/64
Full Address:   2a03:2880:0020:4f06:face:b00c:0000:0014
Address:        2a03:2880:20:4f06:face:b00c:0:14
Full Network:   2a03:2880:0020:4f06:0000:0000:0000:0000/64
Network:        2a03:2880:20:4f06::/64
Netmask:        ffff:ffff:ffff:ffff:: = 64
Reverse DNS:    6.0.f.4.0.2.0.0.0.8.8.2.3.0.a.2.ip6.arpa.

Address space:  Global Unicast
HostMin:        2a03:2880:20:4f06::
HostMax:        2a03:2880:20:4f06:ffff:ffff:ffff:ffff
Hosts/Net:      2^(64) = 18446744073709551616

Country code:   IE
Country:        Ireland
```

```
$ ./ipcalc -pnmb --minaddr --maxaddr --addrspace --geoinfo 2a03:2880:20:4f06:face:b00c:0:14/64
NETMASK=ffff:ffff:ffff:ffff::
PREFIX=64
NETWORK=2a03:2880:20:4f06::
MINADDR=2a03:2880:20:4f06::
MAXADDR=2a03:2880:20:4f06:ffff:ffff:ffff:ffff
ADDRSPACE="Global Unicast"
COUNTRY="Ireland"
```



NEWS
* Version 0.2.5 (unreleased)
- Fixed compilation with no geoinformation
- Fixed segmentation faul when compiled with, but run without libmaxminddb.


* Version 0.2.4 (released 2018-07-23)
- Added support for libmaxmind DB API as an alternative to libgeoip.


* Version 0.2.3 (released 2018-05-26)
- Fixed --no-decorate option with --split and IPv6 addresses.
- Fixed build with USE_GEOIP=yes and USE_DYN_GEOIP=no (David Guglielmi)
- Fixed build with libgeoip < 1.6.3


* Version 0.2.2 (released 2017-12-27)
- Added a more silent version of --split when combined with
  --no-decorate option.


* Version 0.2.0 (released 2017-05-15)
- Added the --split (-S) option. This allows splitting networks
  into smaller chunks.


* Version 0.1.8 (released 2016-06-14)
- Do not print the reverse DNS information by default. Only
  when --all-info is specified.


* Version 0.1.7 (released 2016-04-01)
- Assume a classless IPv4 range by default. This breaks compatibility
  with previous versions since, unless --class-prefix is specified,
  the default IPv4 prefix is now /32.
- Added support for printing reverse DNS addresses.


* Version 0.1.6 (released 2015-12-08)
- Introduced color printing.


* Version 0.1.5 (released 2015-11-05)
- Auto-detect IPv6 addresses when generating random networks.
- Removed libpopt dependency by using getopt_long().


* Version 0.1.4 (released 2015-10-14)
- Added the --all-info option to enable the printing of all
  available information.


* Version 0.1.3 (released 2015-10-05)
- Print the country code in informational mode.


* Version 0.1.2 (released 2015-06-09)
- Print the country code when --geoinfo is requested.
- Only quote the strings if there are spaces in them.


* Version 0.1.1 (released 2015-05-19)
- Allow a prefix of 0, or 0.0.0.0 to be compatible with previous
  tool.


* Version 0.1.0 (released 2015-05-18)
- Added --random-private parameter to generate private networks.
- IPv4 and IPv6 addresses are now automatically being detected,
  and the --ipv4 or --ipv6 options are not mandatory to specify.
- Added the --minaddr and --maxaddr options to print the minimum
  and maximum addresses in the provided network.
- Added the --addresses option to print the number of addresses
  in the provided network.
- Added the --addrspace option to print the type of the network
  the address resides on.
- Added the --geoinfo option to print geographic information about
  the given IP address. That will work if libGeoIP is found on
  the system.
- By default, when no parameters are given, general information
  about the provided network/address are printed (also as --info
  option).



netsplit.c
/*
 * Copyright (c) 2003-2016  Simon Ekstrand
 * Copyright (c) 2010-2016  Joachim Nilsson
 * Copyright (c) 2016 Nikos Mavrogiannopoulos
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *  
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdint.h>
#include <sys/socket.h>

#include "ipcalc.h"

#if defined(__FreeBSD__)
#ifndef s6_addr16
#define s6_addr16 __u6_addr.__u6_addr16
#endif
#ifndef s6_addr32
#define s6_addr32 __u6_addr.__u6_addr32
#endif
#endif

static const char *numtoquad(uint32_t num)
{
	static char quad[64];

	num = htonl(num);
	return inet_ntop(AF_INET, &num, quad, sizeof(quad));
}

void show_split_networks_v4(unsigned split_prefix, const struct ip_info_st *info, unsigned flags)
{
	char buf[64];
	uint32_t diff, start, end;
	size_t maxlen = 0;
	unsigned count;
	uint32_t splitmask = ntohl(prefix2mask(split_prefix));
	uint32_t nmask = ntohl(prefix2mask(info->prefix));
	struct in_addr net, broadcast;

	if (splitmask < nmask) {
		if (!beSilent)
			fprintf(stderr, "Cannot subnet to /%d with this base network, use a prefix > /%d\n",
				split_prefix, info->prefix);
		exit(1);
	}

	if (!(flags & FLAG_NO_DECORATE))
		printf("[Split networks]\n");

	if (inet_pton(AF_INET, info->network, &net) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv4 address: %s\n", info->network);
		exit(1);
	}
	net.s_addr = ntohl(net.s_addr);

	if (inet_pton(AF_INET, info->broadcast, &broadcast) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad broadcast address: %s\n", info->broadcast);
		exit(1);
	}
	broadcast.s_addr = ntohl(broadcast.s_addr);

	diff  = 0xffffffff - splitmask + 1;
	start = net.s_addr;
	end   = net.s_addr + diff - 1;

	/* Figure out max width of a network string. */
	while (1) {
		size_t len;

		len = snprintf(buf, sizeof(buf), "%s", numtoquad(start));
		if (len > maxlen)
			maxlen = len;

		start += diff;
		if (end == 0xffffffff || end >= broadcast.s_addr)
			break;
		end += diff;
	}

	start = net.s_addr;
	end = net.s_addr + diff - 1;
	count = 0;
	while (1) {
		if (!(flags & FLAG_NO_DECORATE))
			default_printf("Network:\t", "%s/%u\n", numtoquad(start), split_prefix);
		else
			printf("%s/%u\n", numtoquad(start), split_prefix);

		count++;
		start += diff;
		if (end == 0xffffffff || end >= broadcast.s_addr)
			break;
		end += diff;
	}

	if (!(flags & FLAG_NO_DECORATE)) {
		dist_printf("\nTotal:  \t", "%u\n", count);
		dist_printf("Hosts/Net:\t", "%s\n", ipv4_prefix_to_hosts(buf, sizeof(buf), split_prefix));
	}
}

static const char *ipv6tostr(struct in6_addr *ip)
{
	static char str[64];

	return inet_ntop(AF_INET6, ip, str, sizeof(str));
}

static void v6add(struct in6_addr *a, const struct in6_addr *b)
{
	int i, j;
	uint32_t tmp;

	for (i = 15; i >= 0; i--) {
		tmp = (uint32_t)a->s6_addr[i] + (uint32_t)b->s6_addr[i];
		if (tmp > 0xff && i > 0) {
			j = i - 1;
			for (j=i-1;j>=0;j--) {
				a->s6_addr[j]++;
				if (a->s6_addr[j] != 0)
					break;
			}
		}

		a->s6_addr[i] = tmp & 0xff;
	}
}

void show_split_networks_v6(unsigned split_prefix, const struct ip_info_st *info, unsigned flags)
{
	int i, j, k;
	unsigned count;
	struct in6_addr splitmask, net, netmask, sdiff, ediff, start, end, tmpaddr, netlast;
	char buf[32];

	if (inet_pton(AF_INET6, info->network, &net) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv6 address: %s\n", info->network);
		exit(1);
	}

	if (inet_pton(AF_INET6, info->hostmax, &netlast) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv6 address: %s\n", info->hostmax);
		exit(1);
	}

	if (inet_pton(AF_INET6, info->netmask, &netmask) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv6 mask: %s\n", info->netmask);
		exit(1);
	}

	if (ipv6_prefix_to_mask(split_prefix, &splitmask) < 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: IPv6 prefix: %d\n", split_prefix);
		exit(1);
	}

	i = 0;
	j = 0;
	do {
		if (splitmask.s6_addr[i] > netmask.s6_addr[i])
			j = 1;
		if (netmask.s6_addr[i] > splitmask.s6_addr[i])
			j = 2;
		i++;
	} while (i < 16 && !j);

	if (j == 2) {
		if (!beSilent)
			fprintf(stderr, "Cannot subnet to /%d with this base network, use a prefix > /%d\n",
				split_prefix, info->prefix);
		exit(1);
	}

	memset(&sdiff, 0, sizeof(sdiff));
	memset(&ediff, 0, sizeof(ediff));

	for (i = 0; i < 16; i++) {
		if (splitmask.s6_addr)
			sdiff.s6_addr[i] = 0xff - splitmask.s6_addr[i];
		end.s6_addr[i] = net.s6_addr[i] + sdiff.s6_addr[i];
	}

	memcpy(&start, &net, sizeof(net));
	memcpy(&ediff, &sdiff, sizeof(sdiff));

	memset(&tmpaddr, 0, sizeof(tmpaddr));
	tmpaddr.s6_addr[15] = 1;
	v6add(&sdiff, &tmpaddr);

	if (!(flags & FLAG_NO_DECORATE))
		printf("[Split networks]\n");

	i = count = 0;
	while (!i) {
		if (!(flags & FLAG_NO_DECORATE))
			default_printf("Network:\t", "%s/%u\n", ipv6tostr(&start), split_prefix);
		else
			printf("%s/%u\n", ipv6tostr(&start), split_prefix);

		v6add(&start, &sdiff);

		j = 0;
		for (k = 0; k < 16; k+=2)
			if (end.s6_addr[k] != 0xff && end.s6_addr[k+1] != 0xff)
				j = 1;
		if (!j)
			i = 1;

		j = 0;
		k = 0;
		do {
			if (end.s6_addr[k] > netlast.s6_addr[k])
				j = 1;
			if (netlast.s6_addr[k] > end.s6_addr[k])
				j = 2;
			k++;
		} while (k < 16 && !j);

		if (!j || j == 1)
			i = 1;

		memset(&end, 0, sizeof(end));

		v6add(&end, &start);
		v6add(&end, &ediff);
		count++;
	}

	if (!(flags & FLAG_NO_DECORATE)) {
		dist_printf("\nTotal:  \t", "%u\n", count);
		dist_printf("Hosts/Net:\t", "%s\n", ipv6_prefix_to_hosts(buf, sizeof(buf), split_prefix));
	}
}



ipcalc-geoip.c
/*
 * Copyright (c) 2015 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authors:
 *   Nikos Mavrogiannopoulos <nmav@redhat.com>
 */

#define _GNU_SOURCE		/* asprintf */
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdarg.h>
#include "ipcalc.h"

#ifdef USE_GEOIP

# include <GeoIP.h>
# include <GeoIPCity.h>

#define GEOIP_SILENCE 16	/* fix libgeoip < 1.6.3 */

# ifdef USE_RUNTIME_LINKING
#  include <dlfcn.h>

typedef void (*_GeoIP_setup_dbfilename_func)(void);
typedef GeoIP * (*GeoIP_open_type_func)(int type, int flags);
typedef const char * (*GeoIP_country_name_by_id_func)(GeoIP * gi, int id);
typedef void (*GeoIP_delete_func)(GeoIP * gi);
typedef GeoIPRecord * (*GeoIP_record_by_ipnum_func)(GeoIP * gi, unsigned long ipnum);
typedef int (*GeoIP_id_by_ipnum_func)(GeoIP * gi, unsigned long ipnum);
typedef int (*GeoIP_id_by_ipnum_v6_func)(GeoIP * gi, geoipv6_t ipnum);
typedef GeoIPRecord *(*GeoIP_record_by_ipnum_v6_func)(GeoIP * gi, geoipv6_t ipnum);
typedef const char *(*GeoIP_code_by_id_func)(int id);

static _GeoIP_setup_dbfilename_func p_GeoIP_setup_dbfilename;
static GeoIP_open_type_func pGeoIP_open_type;
static GeoIP_country_name_by_id_func pGeoIP_country_name_by_id;
static GeoIP_code_by_id_func pGeoIP_code_by_id;
static GeoIP_delete_func pGeoIP_delete;
static GeoIP_record_by_ipnum_func pGeoIP_record_by_ipnum;
static GeoIP_id_by_ipnum_func pGeoIP_id_by_ipnum;
static GeoIP_id_by_ipnum_v6_func pGeoIP_id_by_ipnum_v6;
static GeoIP_record_by_ipnum_v6_func pGeoIP_record_by_ipnum_v6;

#define LIBNAME LIBPATH"/libGeoIP.so.1"

int geo_setup(void)
{
	static void *ld = NULL;
	static int ret = 0;
	static char err[256] = {0};

	if (ld != NULL || ret != 0) {
	    	if (!beSilent && err[0] != 0) {
	    		fprintf(stderr, "%s", err);
		}
		return ret;
	}

	ld = dlopen(LIBNAME, RTLD_LAZY);
	if (ld == NULL) {
		snprintf(err, sizeof(err), "ipcalc: could not open %s\n", LIBNAME);
		ret = -1;
		goto exit;
	}

	p_GeoIP_setup_dbfilename = dlsym(ld, "_GeoIP_setup_dbfilename");

	pGeoIP_open_type = dlsym(ld, "GeoIP_open_type");
	pGeoIP_country_name_by_id = dlsym(ld, "GeoIP_country_name_by_id");
	pGeoIP_delete = dlsym(ld, "GeoIP_delete");
	pGeoIP_record_by_ipnum = dlsym(ld, "GeoIP_record_by_ipnum");
	pGeoIP_id_by_ipnum = dlsym(ld, "GeoIP_id_by_ipnum");
	pGeoIP_id_by_ipnum_v6 = dlsym(ld, "GeoIP_id_by_ipnum_v6");
	pGeoIP_record_by_ipnum_v6 = dlsym(ld, "GeoIP_record_by_ipnum_v6");
	pGeoIP_code_by_id = dlsym(ld, "GeoIP_code_by_id");

	if (pGeoIP_open_type == NULL || pGeoIP_country_name_by_id == NULL ||
	    pGeoIP_delete == NULL || pGeoIP_record_by_ipnum == NULL ||
	    pGeoIP_id_by_ipnum == NULL || pGeoIP_id_by_ipnum_v6 == NULL ||
	    pGeoIP_record_by_ipnum_v6 == NULL) {
		snprintf(err, sizeof(err), "ipcalc: could not find symbols in libGeoIP\n");
	    	ret = -1;
	    	goto exit;
	}

	ret = 0;
 exit:
	return ret;
}

# else

extern void _GeoIP_setup_dbfilename(void);
#  define p_GeoIP_setup_dbfilename _GeoIP_setup_dbfilename
#  define pGeoIP_open_type GeoIP_open_type
#  define pGeoIP_country_name_by_id GeoIP_country_name_by_id
#  define pGeoIP_delete GeoIP_delete
#  define pGeoIP_record_by_ipnum GeoIP_record_by_ipnum
#  define pGeoIP_id_by_ipnum GeoIP_id_by_ipnum
#  define pGeoIP_id_by_ipnum_v6 GeoIP_id_by_ipnum_v6
#  define pGeoIP_record_by_ipnum_v6 GeoIP_record_by_ipnum_v6
#  define pGeoIP_code_by_id GeoIP_code_by_id
# endif

static void geo_ipv4_lookup(struct in_addr ip, char **country, char **ccode, char **city, char **coord)
{
	GeoIP *gi;
	GeoIPRecord *gir;
	int country_id;
	const char *p;

	if (geo_setup() != 0)
		return;

	ip.s_addr = ntohl(ip.s_addr);

	p_GeoIP_setup_dbfilename();

	gi = pGeoIP_open_type(GEOIP_COUNTRY_EDITION, GEOIP_STANDARD | GEOIP_SILENCE);
	if (gi != NULL) {
		gi->charset = GEOIP_CHARSET_UTF8;

		country_id = pGeoIP_id_by_ipnum(gi, ip.s_addr);
		if (country_id < 0) {
			return;
		}
		p = pGeoIP_country_name_by_id(gi, country_id);
		if (p)
			*country = safe_strdup(p);

		p = pGeoIP_code_by_id(country_id);
		if (p)
			*ccode = safe_strdup(p);

		pGeoIP_delete(gi);
	}

	gi = pGeoIP_open_type(GEOIP_CITY_EDITION_REV1, GEOIP_STANDARD | GEOIP_SILENCE);
	if (gi != NULL) {
		gi->charset = GEOIP_CHARSET_UTF8;

		gir = pGeoIP_record_by_ipnum(gi, ip.s_addr);

		if (gir && gir->city)
			*city = safe_strdup(gir->city);

		if (gir && gir->longitude != 0 && gir->longitude != 0)
			safe_asprintf(coord, "%f,%f", gir->latitude, gir->longitude);

		pGeoIP_delete(gi);
	} else {
		gi = pGeoIP_open_type(GEOIP_CITY_EDITION_REV0, GEOIP_STANDARD | GEOIP_SILENCE);
		if (gi != NULL) {
			gi->charset = GEOIP_CHARSET_UTF8;

			gir = pGeoIP_record_by_ipnum(gi, ip.s_addr);

			if (gir && gir->city)
				*city = safe_strdup(gir->city);

			if (gir && gir->longitude != 0 && gir->longitude != 0)
				safe_asprintf(coord, "%f,%f", gir->latitude, gir->longitude);

			pGeoIP_delete(gi);
		}
	}

	return;
}

static void geo_ipv6_lookup(struct in6_addr *ip, char **country, char **ccode, char **city, char **coord)
{
	GeoIP *gi;
	GeoIPRecord *gir;
	int country_id;
	const char *p;

	if (geo_setup() != 0)
		return;

	p_GeoIP_setup_dbfilename();

	gi = pGeoIP_open_type(GEOIP_COUNTRY_EDITION_V6, GEOIP_STANDARD | GEOIP_SILENCE);
	if (gi != NULL) {
		gi->charset = GEOIP_CHARSET_UTF8;

		country_id = pGeoIP_id_by_ipnum_v6(gi, (geoipv6_t)*ip);
		if (country_id < 0) {
			return;
		}
		p = pGeoIP_country_name_by_id(gi, country_id);
		if (p)
			*country = safe_strdup(p);

		p = pGeoIP_code_by_id(country_id);
		if (p)
			*ccode = safe_strdup(p);

		pGeoIP_delete(gi);
	}

	gi = pGeoIP_open_type(GEOIP_CITY_EDITION_REV1_V6, GEOIP_STANDARD | GEOIP_SILENCE);
	if (gi != NULL) {
		gi->charset = GEOIP_CHARSET_UTF8;

		gir = pGeoIP_record_by_ipnum_v6(gi, (geoipv6_t)*ip);

		if (gir && gir->city)
			*city = safe_strdup(gir->city);

		if (gir && gir->longitude != 0 && gir->longitude != 0)
			safe_asprintf(coord, "%f,%f", gir->latitude, gir->longitude);

		pGeoIP_delete(gi);
	} else {
		gi = pGeoIP_open_type(GEOIP_CITY_EDITION_REV0_V6, GEOIP_STANDARD | GEOIP_SILENCE);
		if (gi != NULL) {
			gi->charset = GEOIP_CHARSET_UTF8;

			gir = pGeoIP_record_by_ipnum_v6(gi, (geoipv6_t)*ip);

			if (gir && gir->city)
				*city = safe_strdup(gir->city);

			if (gir && gir->longitude != 0 && gir->longitude != 0)
				safe_asprintf(coord, "%f,%f", gir->latitude, gir->longitude);

			pGeoIP_delete(gi);
		}
	}

	return;
}

void geo_ip_lookup(const char *ip, char **country, char **ccode, char **city, char **coord)
{
        struct in_addr ipv4;
        struct in6_addr ipv6;
        if (inet_pton(AF_INET, ip, &ipv4) == 1) {
              geo_ipv4_lookup(ipv4, country, ccode, city, coord);
        } else if (inet_pton(AF_INET6, ip, &ipv6) == 1) {
              geo_ipv6_lookup(&ipv6, country, ccode, city, coord);
        }
        return;
}

#endif


ipcalc-reverse.c
/*
 * Copyright (c) 2015 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authors:
 *   Nikos Mavrogiannopoulos <nmav@redhat.com>
 */

#define _GNU_SOURCE		/* asprintf */
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdarg.h>
#include "ipcalc.h"

/* draft-ietf-dnsop-rfc2317bis-00 replaces that legacy style
 */
#undef USE_RFC2317_STYLE

char *calc_reverse_dns4(struct in_addr ip, unsigned prefix, struct in_addr network, struct in_addr broadcast)
{
	char *str = NULL;
	int ret = -1;

	unsigned byte1 = (ntohl(ip.s_addr) >> 24) & 0xff;
	unsigned byte2 = (ntohl(ip.s_addr) >> 16) & 0xff;
	unsigned byte3 = (ntohl(ip.s_addr) >> 8) & 0xff;
	unsigned byte4 = (ntohl(ip.s_addr)) & 0xff;

#ifdef USE_RFC2317_STYLE
	if (prefix == 32) {
		ret = safe_asprintf(&str, "%u.%u.%u.%u.in-addr.arpa.", byte4, byte3, byte2, byte1);
	} else if (prefix == 24) {
		ret = safe_asprintf(&str, "%u.%u.%u.in-addr.arpa.", byte3, byte2, byte1);
	} else if (prefix == 16) {
		ret = safe_asprintf(&str, "%u.%u.in-addr.arpa.", byte2, byte1);
	} else if (prefix == 8) {
		ret = safe_asprintf(&str, "%u.in-addr.arpa.", byte1);
	} else if (prefix > 24) {
		ret = safe_asprintf(&str, "%u/%u.%u.%u.%u.in-addr.arpa.", byte4, prefix, byte3, byte2, byte1);
	} else if (prefix > 16) {
		ret = safe_asprintf(&str, "%u/%u.%u.%u.in-addr.arpa.", byte3, prefix, byte2, byte1);
	} else if (prefix > 8) {
		ret = safe_asprintf(&str, "%u/%u.%u.in-addr.arpa.", byte2, prefix, byte1);
	}
#else
	if (prefix == 32) {
		ret = safe_asprintf(&str, "%u.%u.%u.%u.in-addr.arpa.", byte4, byte3, byte2, byte1);
	} else if (prefix == 24) {
		ret = safe_asprintf(&str, "%u.%u.%u.in-addr.arpa.", byte3, byte2, byte1);
	} else if (prefix == 16) {
		ret = safe_asprintf(&str, "%u.%u.in-addr.arpa.", byte2, byte1);
	} else if (prefix == 8) {
		ret = safe_asprintf(&str, "%u.in-addr.arpa.", byte1);
	} else if (prefix > 24) {
		unsigned min = (ntohl(network.s_addr)) & 0xff;
		unsigned max = (ntohl(broadcast.s_addr)) & 0xff;
		ret = safe_asprintf(&str, "%u-%u.%u.%u.%u.in-addr.arpa.", min, max, byte3, byte2, byte1);
	} else if (prefix > 16) {
		unsigned min = (ntohl(network.s_addr) >> 8) & 0xff;
		unsigned max = (ntohl(broadcast.s_addr) >> 8) & 0xff;
		ret = safe_asprintf(&str, "%u-%u.%u.%u.in-addr.arpa.", min, max, byte2, byte1);
	} else if (prefix > 8) {
		unsigned min = (ntohl(network.s_addr) >> 16) & 0xff;
		unsigned max = (ntohl(broadcast.s_addr) >> 16) & 0xff;
		ret = safe_asprintf(&str, "%u-%u.%u.in-addr.arpa.", min, max, byte1);
	}
#endif

	if (ret == -1)
		return NULL;
	return str;
}

static char hexchar(unsigned int val)
{
	if (val < 10)
		return '0' + val;
	if (val < 16)
		return 'a' + val - 10;
	abort();
}

char *calc_reverse_dns6(struct in6_addr *ip, unsigned prefix)
{
	unsigned i, j = 0;
	char str[256];
	unsigned max = prefix / 8;

	if (prefix % 4 != 0)
		return NULL;

	if (prefix % 8 == 4) {
		str[j++] = hexchar(ip->s6_addr[(prefix + 4) / 8 - 1] >> 4);
		str[j++] = '.';
	}

	for (i = 0; i < max; i++) {
		str[j++] = hexchar(ip->s6_addr[max - 1 - i] & 0xf);
		str[j++] = '.';

		str[j++] = hexchar(ip->s6_addr[max - 1 - i] >> 4);
		str[j++] = '.';

	}

	strcpy(&str[j], "ip6.arpa.");

	return safe_strdup(str);
}


ipcalc.h
/*
 * Copyright (c) 2016 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authors:
 *   Nikos Mavrogiannopoulos <nmav@redhat.com>
 */

#ifndef _IPCALC_H
#define _IPCALC_H

int __attribute__((__format__(printf, 2, 3))) safe_asprintf(char **strp, const char *fmt, ...);

#if defined(USE_GEOIP)
  void geo_ip_lookup(const char *ip, char **country, char **ccode, char **city, char  **coord);
  int geo_setup(void);
# ifndef USE_RUNTIME_LINKING
#   define geo_setup() 0
# endif
#elif defined(USE_MAXMIND)
  void geo_ip_lookup(const char *ip, char **country, char **ccode, char **city, char  **coord);
  int geo_setup(void);
# ifndef USE_RUNTIME_LINKING
#   define geo_setup() 0
# endif
#else
# define geo_ipv4_lookup(x,y,z,w,a)
# define geo_ipv6_lookup(x,y,z,w,a)
# define geo_setup() -1
#endif

char __attribute__((warn_unused_result)) *safe_strdup(const char *str);

char *calc_reverse_dns4(struct in_addr ip, unsigned prefix, struct in_addr net, struct in_addr bcast);
char *calc_reverse_dns6(struct in6_addr *ip, unsigned prefix);

uint32_t prefix2mask(int prefix);
int ipv6_prefix_to_mask(unsigned prefix, struct in6_addr *mask); 

char *ipv4_prefix_to_hosts(char *hosts, unsigned hosts_size, unsigned prefix);
char *ipv6_prefix_to_hosts(char *hosts, unsigned hosts_size, unsigned prefix);

typedef struct ip_info_st {
	char *ip;
	char *expanded_ip;
	char *expanded_network;
	char *reverse_dns;

	char *network;
	char *broadcast;	/* ipv4 only */
	char *netmask;
	char *hostname;
	char *geoip_country;
	char *geoip_ccode;
	char *geoip_city;
	char *geoip_coord;
	char hosts[64];		/* number of hosts in text */
	unsigned prefix;

	char *hostmin;
	char *hostmax;
	const char *type;
	const char *class;
} ip_info_st;

#define FLAG_RESOLVE_HOST 1
#define FLAG_RESOLVE_IP (1<<1)
#define FLAG_CHECK_ADDRESS (1<<2)
#define FLAG_SHOW_INFO (1<<3)
#define FLAG_SHOW_BROADCAST (1<<6)
#define FLAG_SHOW_NETMASK (1<<7)
#define FLAG_SHOW_NETWORK (1<<8)
#define FLAG_SHOW_PREFIX (1<<9)
#define FLAG_SHOW_MINADDR (1<<10)
#define FLAG_SHOW_MAXADDR (1<<11)
#define FLAG_SHOW_ADDRESSES (1<<12)
#define FLAG_SHOW_ADDRSPACE (1<<13)
#define FLAG_GET_GEOIP (1<<14)
#define FLAG_SHOW_GEOIP ((1<<15)|FLAG_GET_GEOIP)
#define FLAG_SHOW_ALL_INFO ((1<<16)|FLAG_SHOW_INFO)
#define FLAG_SHOW_REVERSE (1<<17)
#define FLAG_ASSUME_CLASS_PREFIX (1<<18)
#define FLAG_SPLIT (1<<19)
#define FLAG_NO_DECORATE (1<<20)

/* Flags that are not real options */
#define FLAGS_TO_IGNORE (FLAG_GET_GEOIP|FLAG_SPLIT|FLAG_NO_DECORATE|FLAG_ASSUME_CLASS_PREFIX|(1<<16))
#define FLAGS_TO_IGNORE_MASK (~FLAGS_TO_IGNORE)

void show_split_networks_v4(unsigned split_prefix, const struct ip_info_st *info, unsigned flags);
void show_split_networks_v6(unsigned split_prefix, const struct ip_info_st *info, unsigned flags);

#define KBLUE  "\x1B[34m"
#define KMAG   "\x1B[35m"
#define KRESET "\033[0m"

#define default_printf(...) color_printf(KBLUE, __VA_ARGS__)
#define dist_printf(...) color_printf(KMAG, __VA_ARGS__)

void
__attribute__ ((format(printf, 3, 4)))
color_printf(const char *color, const char *title, const char *fmt, ...);

extern int beSilent;

#endif
